eskip/json.go

```diff
@@ -57,28 +75,147 @@ func marshalJsonPredicates(r *Route) []*Predicate {
 }
 
+func extractStringArgs(args []interface{}, count int) ([]string, error) {
+	if count > 0 && len(args) != count {
+		return nil, ErrInvalidArgCount
+	}
+
+	if count <= 0 {
+		count = len(args)
+	}
+
+	sargs := make([]string, count)
+	for i := range args {
+		sa, ok := args[i].(string)
+		if !ok {
+			return nil, ErrInvalidArgType
+		}
+
+		sargs[i] = sa
+	}
+
+	return sargs, nil
+}
+
+func unmarshalJSONPredicates(r *Route, p []*Predicate) error {
+	var ps []*Predicate
+	for _, pi := range p {
+		switch pi.Name {
+		case "Method":
+			a, err := extractStringArgs(pi.Args, 1)
+			if err != nil {
+				return err
+			}
+
+			r.Method = a[0]
+		case "Path":
+			a, err := extractStringArgs(pi.Args, 1)
+			if err != nil {
+				return err
+			}
+
+			r.Path = a[0]
+		case "HostRegexp":
+			a, err := extractStringArgs(pi.Args, 0)
+			if err != nil {
+				return err
+			}
+
+			r.HostRegexps = a
+		case "PathRegexp":
+			a, err := extractStringArgs(pi.Args, 0)
+			if err != nil {
+				return err
+			}
+
+			r.PathRegexps = a
+		case "Header":
+			a, err := extractStringArgs(pi.Args, 2)
+			if err != nil {
+				return err
+			}
+
+			if r.Headers == nil {
+				r.Headers = make(map[string]string)
+			}
+
+			r.Headers[a[0]] = a[1]
+		case "HeaderRegexp":
+			a, err := extractStringArgs(pi.Args, 2)
+			if err != nil {
+				return err
+			}
+
+			if r.HeaderRegexps == nil {
+				r.HeaderRegexps = make(map[string][]string)
+			}
+
+			r.HeaderRegexps[a[0]] = append(r.HeaderRegexps[a[0]], a[1])
+		default:
+			ps = append(ps, pi)
+		}
+	}
+
+	r.Predicates = ps
+	return nil
+}
+
 func marshalNameArgs(name string, args []interface{}) ([]byte, error) {
 	if args == nil {
 		args = []interface{}{}
 	}
 
-	return json.Marshal(&struct {
-		Name string        `json:"name"`
-		Args []interface{} `json:"args"`
-	}{
+	var buf bytes.Buffer
+	e := json.NewEncoder(&buf)
+	e.SetEscapeHTML(false) // why not working :(
+
+	err := e.Encode(&jsonNameArgs{
 		Name: name,
 		Args: args,
 	})
+
+	return buf.Bytes(), err
+}
+
+func unmarshalNameArgs(b []byte) (string, []interface{}, error) {
+	var jna jsonNameArgs
+	err := json.Unmarshal(b, &jna)
+	return jna.Name, jna.Args, err
 }
 
+// MarshalJSON returns the JSON representation of a filter.
+//
+// (Typically used only as part of the unmarshalling complete routes.)
 func (f *Filter) MarshalJSON() ([]byte, error) {
 	return marshalNameArgs(f.Name, f.Args)
 }
 
+// UnmarshalJSON returns the JSON representation of a filter.
+//
+// (Typically used only as part of the unmarshalling complete routes.)
+func (f *Filter) UnmarshalJSON(b []byte) error {
+	var err error
+	f.Name, f.Args, err = unmarshalNameArgs(b)
+	return err
+}
+
+// MarshalJSON returns the JSON representation of a predicate.
+//
+// (Typically used only as part of the unmarshalling complete routes.)
 func (p *Predicate) MarshalJSON() ([]byte, error) {
 	return marshalNameArgs(p.Name, p.Args)
 }
 
+// UnmarshalJSON returns the JSON representation of a predicate.
```

Comment:
Are you sure that you don’t do it the other way around?
Marshal and unmarshal have very similar doc strings.

Reply:
I agree that the doc comment should rather say something like this:

"UnmarshalJSON parses a predicate object from its JSON representation."